### PR TITLE
Charm: provide MPI

### PR DESCRIPTION
**Pull request by matthiasdiener**

### CODE REVIEW

File: var/spack/repos/builtin/packages/charm/package.py [hunk 1]
```diff
@@ -92,6 +92,24 @@ class Charm(Package):
     variant("production", default=True, description="Build charm++ with all optimizations")
     variant("tracing", default=False, description="Enable tracing modules")
 
+    # FIXME: backend=mpi also provides mpi, but this can't be expressed in
+    # spack currently
+    # FIXME: shorten this to 'build-target!=charm++'
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/charm/package.py]: You could shorten it quite a bit with a for-loop.

Comment (matthiasdiener) [reply var/spack/repos/builtin/packages/charm/package.py]: Done, thanks! Do you have a suggestion on how to fix the `backend=mpi` situation? Currently, this results in an error message `==> Error: maximum recursion depth exceeded`, since in this case we have `depends_on('mpi')` as well as `provides('mpi')`.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/charm/package.py]: No idea.

Comment (matthiasdiener) [reply var/spack/repos/builtin/packages/charm/package.py]: Ok. Not having this for the mpi backend might be enough for now I think.

**Comment by adamjstewart** [comment]
I don't know much about Charm++, so pinging others who have contributed to the package in the past to help review this PR: @jimmy04500 @junghans @eschnett

**Comment by matthiasdiener** [comment]
I think this is ready for review.

File: var/spack/repos/builtin/packages/charm/package.py [hunk 2]
```diff
@@ -92,6 +92,25 @@ class Charm(Package):
     variant("production", default=True, description="Build charm++ with all optimizations")
     variant("tracing", default=False, description="Enable tracing modules")
 
+    # FIXME: backend=mpi also provides mpi, but this can't be expressed in
+    # spack currently
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/charm/package.py]: Can you make this more specific?  I think you're saying it can't be expressed b/c there would be two MPI providers in that configuration, and we don't allow two providers of the same virtual dependency in the same DAG.  Is that right?

Just for my own edification, how *do* you handle that case?  If the backend is MPI and Charm++ is providing MPI, how do you avoid symbol conflicts at runtime?

Comment (matthiasdiener) [reply var/spack/repos/builtin/packages/charm/package.py]: I tried to clarify the comment a bit. The error message I get when writing `depends_on("mpi")` and `provides("mpi")` in the package is `==> Error: Cannot depend on 'charm@6.8.2%gcc@7.3.0 backend=mpi build-target=LIBS ~cuda~omp~papi+production~pthreads+shared+smp~syncft~tcp~tracing arch=linux-ubuntu18.04-x86_64' twice`.

In Charm++/AMPI, we rename symbols from `MPI_Foo` to `AMPI_Foo` in the user application when using MPI as the backend, which avoids the symbol conflict.

**Comment by tgamblin** [comment]
@matthiasdiener: Thanks!  That's a Spack issue at the moment but it would be interesting to figure out how we could add better support for `AMPI`.  @bhatele might be interested in this.

Minor note: the `build-target` variant is a little mechanical-sounding.  It seems to me (though I'm not an expert) that a user would always want to build the libraries and *usually* would want to build `charm++`.  `ampi` seems like something another package would want to depend on, or that a user would specify on the command line.  e.g.:

```console
spack install my-physics-code+mpi ^charm+ampi
```

but currently you have to write:

```console
spack install my-physics-code+mpi ^charm build-target=ampi
```

which is ok, but I have to think about it more, and I have to think about the notion of "build targets" rather than "charm has a feature called ampi".

Anyway, just a thought.

**Comment by yellowhat** [comment]
Hi,
I am still having this issue:
```
spack spec namd ^charmpp backend=mpi
Input spec
--------------------------------
namd
    ^charmpp backend=mpi

Concretized
--------------------------------
==> Error: maximum recursion depth exceeded
```

**Comment by scheibelp** [comment]
@yellowhat I don't think`backend=mpi` is supported yet (Spack's concretizer is not yet sophisticated enough to handle it)

**Comment by scheibelp** [comment]
@tgamblin FYI I think this is a distinct use case (a package that provides a virtual but also depends on that same virtual in some cases) so I have added the`concretizer-use-case` tag
